config/v1/types_cluster_monitoring.go

```diff
@@ -0,0 +1,84 @@
+/*
+Copyright 2024.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// Compatibility level 4: No compatibility is provided, the API can change at any point for any reason. These capabilities should not be used by applications needing long term support.
+// +openshift:compatibility-gen:internal
+// +openshift:api-approved.openshift.io=https://github.com/openshift/api/pull/1929
+// +openshift:file-pattern=cvoRunLevel=0000_10,operatorName=config-operator,operatorOrdering=01
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:path=clustermonitoring,scope=Cluster
+// +kubebuilder:subresource:status
+// +kubebuilder:metadata:annotations=release.openshift.io/bootstrap-required=true
+// +kubebuilder:metadata:annotations="description=Cluster Monitoring Operators configuration API"
+// ClusterMonitoring is the Schema for the Cluster Monitoring Operators API
+type ClusterMonitoring struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// spec holds user configuration for the Cluster Monitoring Operator
+	// +kubebuilder:validation:Required
+	// +required
+	Spec ClusterMonitoringSpec `json:"spec,omitempty"`
+	// status holds observed values from the cluster. They may not be overridden.
+	// +optional
+	Status ClusterMonitoringStatus `json:"status,omitempty"`
+}
+
+// MonitoringOperatorStatus defines the observed state of MonitoringOperator
+type ClusterMonitoringStatus struct {
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// Compatibility level 4: No compatibility is provided, the API can change at any point for any reason. These capabilities should not be used by applications needing long term support.
+// +openshift:compatibility-gen:internal
+type ClusterMonitoringList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []ClusterMonitoring `json:"items"`
+}
+
+// MonitoringOperatorSpec defines the desired state of Cluster Monitoring Operator
+type ClusterMonitoringSpec struct {
+	// `UserWorkload` set the deployment mode for user-defined monitoring.
+	// +kubebuilder:validation:Optional
+	// +optional
+	UserDefined UserDefinedMonitoring `json:"userDefinedMonitoring,omitempty"`
```

Comment:
Apologies for missing this before: With `UserDefined` being optional we currently essentially have three scenarios: `UserDefined` is absent, set to `Disabled` or set to `SingleNamespace`.
Unless I'm missing something, the first two are equivalent, we should probably have one way to disable UWM.
I seem to remember we went with a `Mode` field since we want to give users the option to disable UWM without deleting their (as of yet nonexistent) UWM config beyond mode. With that line of thinking we should maybe make this required and default to `{Mode: "Disabled"}`?

Reply:
Yep I think best idea is maybe do it `required`  and default `Disabled`, just in configmap is "optional" and "optional" means "disabled", for that reason I marked it as optional. After that we have extended the options...

```
	// `UserWorkloadEnabled` is a Boolean flag that enables monitoring for user-defined projects.
	UserWorkloadEnabled *bool `json:"enableUserWorkload,omitempty"`
```

Reply:
https://github.com/kubernetes-sigs/controller-tools/issues/893#issuecomment-1991256368

Reply:
Right, I think ran into this before. I guess `optional` plus a default disabled it is then? :) 
Mostly I want to avoid having two ways of disabling UWM (either omit the config _or_ set `mode: disabled`).

Reply:
Yep I am on it but facing some issues with tests

Reply:
When added required [config/v1/tests/clustermonitoring.config.openshift.io/AAA_ungated.yaml](url) has to be changed manually... that was the issue